### PR TITLE
Added new "Plugin metadata" section in docs

### LINKED ISSUE #9413: Document the metadata format
Extracted from #6642.

Once we have a consensus about the metadata format we should document it.

> Document the format in the Framework docs (e.g. the "Contributing" section) and cross-link it at least at the end of "Creating a simple plugin" and maybe the tutorials.

### PR DESCRIPTION
**Pull request by psmyrek**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Docs: Added new "Plugin metadata" section. Closes #9413.

---

### Additional information

I added this new "Plugin metadata" section between "Code style" and "Git commit message convention":

plugin-metadata-in-docs

### CODE REVIEW

File: docs/framework/guides/contributing/plugin-metadata.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+	* `classes` &ndash; CSS class names (a single one or an array of these) that may be applied to the HTML elements defined in the `elements` property.
+	* `styles` &ndash; Inline CSS styles (a single one or an array of these) that may be applied to the HTML elements defined in the `elements` property.
+	* `attributes` &ndash; HTML attributes (a single one or an array of these) other than `class` and `styles` (covered separately), that might be applied to the HTML elements defined in the `elements` property.
+	* `implements` &ndash; A name of an element or a pseudo-element which indicates that HTML elements defined in the `elements` property may contain classes, styles or attributes that are created by other plugins. For example, `implements` equal to `$block` means, that HTML elements may contain classes, styles or attributes that are defined by another plugin, which have `elements` equal to `$block`.
+	* `_comment` &ndash; A human-readable description to explain more complicated cases, for example: the conditions when a given element, class or style can be created.
```
Comment (ma2ciek) [inline docs/framework/guides/contributing/plugin-metadata.md:33]: I'd add a link to one of the metadata files that are provided by us at the bottom of this section.

File: docs/framework/guides/creating-simple-plugin.md [hunk 2]
```diff
@@ -295,4 +295,6 @@ If you would like to read more tutorials, check out the following one:
 * {@link framework/guides/tutorials/implementing-a-block-widget Implementing a block widget}
 * {@link framework/guides/tutorials/implementing-an-inline-widget Implementing an inline widget}
 
-If you are more into reading about the CKEditor 5 architecture, check out the {@link framework/guides/architecture/intro Introduction to CKEditor 5 architecture}.
+If you want to read more about the CKEditor 5 architecture, check out the {@link framework/guides/architecture/intro Introduction to CKEditor 5 architecture}.
+
+If you would like know, how to check what kind of HTML output our plugins return or what UI components are exposed by them, check out the {@link framework/guides/contributing/plugin-metadata plugin metadata} guide.
```
Comment (ma2ciek) [inline docs/framework/guides/creating-simple-plugin.md:300]: This will be rather done in this guide - #9414.

In this guide (creating a simple plugin) we should mention here that all 3rd-party CKEditor 5 packages can document their plugins using this meta file. This will allow for easier integrations with their plugins.

File: docs/framework/guides/contributing/plugin-metadata.md [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+* `path` &ndash; A path to the file, relative to the metadata file that exports the plugin.
+* `className` &ndash; The name of the class used to create the plugin.
+* `requires`&ndash; An array of the plugin's soft requirements and other non-explicit requirements.
+* `registeredToolbars`&ndash; An array of all toolbar names registered by the plugin. These names need to contain the full configuration path (e.g. `table.contentToolbar` and `table.tableToolbar` for `Table` plugin).
+* `uiComponents` &ndash; An array of objects, that describe UI components exported by the plugin. Each object in this array may contain:
+	* `name` &ndash; A name of the component the plugin exports.
+	* `type` &ndash; Component type: `Button`, `SplitButton` or `Dropdown`.
+	* `iconPath` &ndash; A path to the SVG icon for `Button` or `SplitButton` components.
+	* `label` &ndash; Text content for `Dropdown` components.
+	* `toolbars`&ndash; An array of toolbar names, a given UI component can be added to. Some features may be added to multiple toolbars.
```
Comment (ma2ciek) [inline docs/framework/guides/contributing/plugin-metadata.md:26]: There are some missing spaces before &ndash; in this file.

File: docs/framework/guides/contributing/plugin-metadata.md [hunk 4]
```diff
@@ -0,0 +1,33 @@
+---
+category: framework-contributing
+order: 40
+---
+
+# Plugin metadata
+
+Most of CKEditor 5 packages contain a separate `ckeditor5-metadata.json` file in their root directory. This file contains the basic information about plugins that are exported by the package. Only the main ones that put the given functionality together are described in `ckeditor5-metadata.json`. For example, the `ckeditor5-metadata.json` in the `ckeditor5-mention` package does not contain the `MentionEditing` and `MentionUI` plugins, but instead it only contains the metadata for the main `Mention` plugin, which is a "glue plugin", that loads the editing and UI features.
```
Comment (ma2ciek) [inline docs/framework/guides/contributing/plugin-metadata.md:8]: This is a nice part of another guide. But in this guide that is located in the `contributing` section we should rather mention that this file can be written by developers that want to make integrations with their plugins easier.

**Review by ma2ciek — CHANGES_REQUESTED** [review]
We should make this guide more contributor-oriented.

File: docs/framework/guides/contributing/plugin-metadata.md (hunk 1 above)
Comment (psmyrek) [reply docs/framework/guides/contributing/plugin-metadata.md:33]: I did it a little differently: I added a link to the metadata file (not working yet, as metadata are not yet on `master` branch) at the top of this guide, so user can open a real sample file and check it while reading about the individual fields in this format. At the bottom of this guide I added a link and information on how user can see other metadata in our other packages.

File: docs/framework/guides/creating-simple-plugin.md (hunk 2 above)
Comment (psmyrek) [reply docs/framework/guides/creating-simple-plugin.md]: I rewrote this sentence - I hope it's better now.

File: docs/framework/guides/contributing/plugin-metadata.md (hunk 3 above)
Comment (psmyrek) [reply docs/framework/guides/contributing/plugin-metadata.md]: Fixed.

File: docs/framework/guides/contributing/plugin-metadata.md (hunk 4 above)
Comment (psmyrek) [reply docs/framework/guides/contributing/plugin-metadata.md]: I rewrote this section a bit. I think that such a basic introduction with a description of what is the metadata, is really needed. I do not know what else I could add here 🙄

File: docs/framework/guides/contributing/plugin-metadata.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+order: 40
+---
+
+# Plugin metadata
+
+The plugin metadata is located in a `ckeditor5-metadata.json` file in the package root directory. The purpose of this file is to simplify the integration and usage of plugins in CKEditor 5, because this file contains the basic information about plugins that are exported by the package. Only the main ones that put the given functionality together are described in `ckeditor5-metadata.json`. For example, the [metadata file](https://github.com/ckeditor/ckeditor5/blob/master/packages/ckeditor5-mention/ckeditor5-metadata.json) in the `ckeditor5-mention` package does not contain the `MentionEditing` and `MentionUI` plugins, but instead it only contains the metadata for the main `Mention` plugin, which is a "glue plugin", that loads the editing and UI features.
+
+## Data format for plugin metadata
+
+The `ckeditor5-metadata.json` file is a JavaScript object that holds a `plugins` array. Each element of this array is an object containing information about a different plugin  delivered by the package. Below is a list of all properties that can be used to define the metadata for a plugin:
```
Comment (ma2ciek) [inline docs/framework/guides/contributing/plugin-metadata.md:12]:
```suggestion
The `ckeditor5-metadata.json` file is a JSON object that holds the `plugins` array. Each element of this array should be an object containing information about a different plugin delivered by the package. Below is a list of all matadata properties that can be used to describe the plugin:
```

File: docs/framework/guides/contributing/plugin-metadata.md [hunk 6]
```diff
@@ -0,0 +1,35 @@
+---
+category: framework-contributing
+order: 40
+---
+
+# Plugin metadata
+
+The plugin metadata is located in a `ckeditor5-metadata.json` file in the package root directory. The purpose of this file is to simplify the integration and usage of plugins in CKEditor 5, because this file contains the basic information about plugins that are exported by the package. Only the main ones that put the given functionality together are described in `ckeditor5-metadata.json`. For example, the [metadata file](https://github.com/ckeditor/ckeditor5/blob/master/packages/ckeditor5-mention/ckeditor5-metadata.json) in the `ckeditor5-mention` package does not contain the `MentionEditing` and `MentionUI` plugins, but instead it only contains the metadata for the main `Mention` plugin, which is a "glue plugin", that loads the editing and UI features.
```
Comment (ma2ciek) [inline docs/framework/guides/contributing/plugin-metadata.md:8]:
```suggestion
The package metadata should be located in the `ckeditor5-metadata.json` file in the package root directory. The purpose of this file is to simplify the integration and usage of plugins in CKEditor 5, because this file contains the basic information about plugins that are exported by the package. Only the main ones that put the given functionality together should bee described in `ckeditor5-metadata.json`. For example, the [metadata file](https://github.com/ckeditor/ckeditor5/blob/master/packages/ckeditor5-mention/ckeditor5-metadata.json) in the `ckeditor5-mention` package does not contain the `MentionEditing` and `MentionUI` plugins, but instead it only contains the metadata for the main `Mention` plugin, which is a "glue plugin", that loads the editing and UI features.
```

File: docs/framework/guides/contributing/plugin-metadata.md (hunk 5 above)
Comment (psmyrek) [reply docs/framework/guides/contributing/plugin-metadata.md]: Done.

File: docs/framework/guides/contributing/plugin-metadata.md (hunk 6 above)
Comment (psmyrek) [reply docs/framework/guides/contributing/plugin-metadata.md]: Done.

File: docs/framework/guides/contributing/package-metadata.md [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+## Data format for plugin metadata
+
+The `ckeditor5-metadata.json` file is a JSON object that holds the `plugins` array. Each element of this array should be an object containing information about a different plugin delivered by the package. Below is a list of all matadata properties that can be used to describe the plugin:
+
+* `name` &ndash; A human-readable name of the plugin.
+* `description` &ndash; A human-readable short description of the plugin.
+* `docs` &ndash; An absolute or relative URL for the plugin's documentation. If this URL is relative, it leads to the CKEditor 5 documentation in [https://ckeditor.com/docs/ckeditor5/latest/](https://ckeditor.com/docs/ckeditor5/latest/).
+* `path` &ndash; A path to the file, relative to the metadata file that exports the plugin.
+* `className` &ndash; The name of the class used to create the plugin.
```
Comment (ma2ciek) [inline docs/framework/guides/contributing/package-metadata.md:18]: \+ This class should be exported from the file using the `export default` syntax.

File: docs/framework/guides/contributing/package-metadata.md [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+* `requires` &ndash; An array of the plugin's soft requirements and other non-explicit requirements.
```
Comment (ma2ciek) [inline docs/framework/guides/contributing/package-metadata.md:19]: \+ It should contain class names of plugins that should be included if this plugin is added.

Comment (ma2ciek) [reply docs/framework/guides/contributing/package-metadata.md:19]: Perhaps it would be safer to add support for `repository:pluginName` in the future.

File: docs/framework/guides/contributing/package-metadata.md [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+* `registeredToolbars` &ndash; An array of all toolbar names registered by the plugin. These names need to contain the full configuration path (e.g. `table.contentToolbar` and `table.tableToolbar` for `Table` plugin).
```
Comment (ma2ciek) [inline docs/framework/guides/contributing/package-metadata.md:20]:
```suggestion
* `registeredToolbars` &ndash; An array of all toolbar names registered by the plugin. These names need to represent the configuration path (e.g. `table.contentToolbar` for `editorConfig.table.contentToolbar` and `table.tableToolbar` for the `editorConfig.tableToolbar` are registered by the `Table` plugin).
```

File: docs/framework/guides/contributing/package-metadata.md [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+* `uiComponents` &ndash; An array of objects, that describe UI components exported by the plugin. Each object in this array may contain:
+	* `name` &ndash; A name of the component the plugin exports.
```
Comment (ma2ciek) [inline docs/framework/guides/contributing/package-metadata.md:22]: It should match the actual UI name registered by the plugin.

File: docs/framework/guides/contributing/package-metadata.md [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+	* `type` &ndash; Component type: `Button`, `SplitButton` or `Dropdown`.
+	* `iconPath` &ndash; A path to the SVG icon for `Button` or `SplitButton` components.
+	* `label` &ndash; Text content for `Dropdown` components.
+	* `toolbars` &ndash; An array of toolbar names, a given UI component can be added to. Some features may be added to multiple toolbars.
```
Comment (ma2ciek) [inline docs/framework/guides/contributing/package-metadata.md:26]:
```suggestion
	* `toolbars` &ndash; An array of toolbar names, a given UI component can be added to. Some UI componenets may be added to multiple toolbars.
```

File: docs/framework/guides/contributing/package-metadata.md [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+
+* `name` &ndash; A human-readable name of the plugin.
+* `description` &ndash; A human-readable short description of the plugin.
+* `docs` &ndash; An absolute or relative URL for the plugin's documentation. If this URL is relative, it leads to the CKEditor 5 documentation in [https://ckeditor.com/docs/ckeditor5/latest/](https://ckeditor.com/docs/ckeditor5/latest/).
+* `path` &ndash; A path to the file, relative to the metadata file that exports the plugin.
+* `className` &ndash; The name of the class used to create the plugin.
+* `requires` &ndash; An array of the plugin's soft requirements and other non-explicit requirements.
+* `registeredToolbars` &ndash; An array of all toolbar names registered by the plugin. These names need to contain the full configuration path (e.g. `table.contentToolbar` and `table.tableToolbar` for `Table` plugin).
+* `uiComponents` &ndash; An array of objects, that describe UI components exported by the plugin. Each object in this array may contain:
```
Comment (ma2ciek) [inline docs/framework/guides/contributing/package-metadata.md:21]:
```suggestion
* `uiComponents` &ndash; An array of objects, that describes UI components exported by the plugin. Each object in this array may contain:
```

File: docs/framework/guides/contributing/package-metadata.md [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+	* `name` &ndash; A name of the component the plugin exports.
+	* `type` &ndash; Component type: `Button`, `SplitButton` or `Dropdown`.
+	* `iconPath` &ndash; A path to the SVG icon for `Button` or `SplitButton` components.
+	* `label` &ndash; Text content for `Dropdown` components.
+	* `toolbars` &ndash; An array of toolbar names, a given UI component can be added to. Some features may be added to multiple toolbars.
+* `htmlOutput` &ndash; An array of objects, that define all possible HTML elements, which can be created by a given plugin. The main property in this object is `elements`. Other properties (e.g. `classes`, `styles`, `attributes`) only apply to items defined in the `elements` property within a given object. Wildcard character `*` is used to mark any value. Full list of all these properties includes:
```
Comment (ma2ciek) [inline docs/framework/guides/contributing/package-metadata.md:27]:
```suggestion
* `htmlOutput` &ndash; An array of objects, that defines all possible HTML elements, which can be created by a given plugin. The main property in this object is `elements`. Other properties (e.g. `classes`, `styles`, `attributes`) only apply to items defined in the `elements` property within a given object. Wildcard character `*` is used to mark any value. Full list of all these properties includes:
```

File: docs/framework/guides/contributing/package-metadata.md [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+	* `elements` &ndash; HTML elements (a single one or an array of these) that are created by the plugin. The pseudo-element `$block` indicates that a given plugin applies classes, styles or attributes (defined in appropriate properties) for all block elements.
```
Comment (ma2ciek) [inline docs/framework/guides/contributing/package-metadata.md:28]: I'd rephrase it as the first sentence is incorrect for the \`$block\`

File: docs/framework/guides/creating-simple-plugin.md [hunk 15]
```diff
@@ -295,4 +295,6 @@ If you would like to read more tutorials, check out the following one:
 * {@link framework/guides/tutorials/implementing-a-block-widget Implementing a block widget}
 * {@link framework/guides/tutorials/implementing-an-inline-widget Implementing an inline widget}
 
-If you are more into reading about the CKEditor 5 architecture, check out the {@link framework/guides/architecture/intro Introduction to CKEditor 5 architecture}.
+If you want to read more about the CKEditor 5 architecture, check out the {@link framework/guides/architecture/intro Introduction to CKEditor 5 architecture}.
+
+If you want your plugin to be easily integrated and used by other developers, try to define a metadata file for the package that contains it, as described in the {@link framework/guides/contributing/package-metadata package metadata} guide.
```
Comment (ma2ciek) [inline docs/framework/guides/creating-simple-plugin.md:300]:
```suggestion
If you want your plugin to be easily integrated by other developers, learn about {@link framework/guides/contributing/package-metadata package metadata file} that can be added to your package.
```

File: docs/framework/guides/contributing/package-metadata.md (hunk 7 above)
Comment (psmyrek) [reply docs/framework/guides/contributing/package-metadata.md]: Done.

File: docs/framework/guides/contributing/package-metadata.md (hunk 8 above)
Comment (psmyrek) [reply docs/framework/guides/contributing/package-metadata.md]: Done.

File: docs/framework/guides/contributing/package-metadata.md (hunk 9 above)
Comment (psmyrek) [reply docs/framework/guides/contributing/package-metadata.md]: Done.
The `tableToolbar` is also inside `editorConfig.table.tableToolbar`.

File: docs/framework/guides/contributing/package-metadata.md (hunk 10 above)
Comment (psmyrek) [reply docs/framework/guides/contributing/package-metadata.md]: Done.

File: docs/framework/guides/contributing/package-metadata.md (hunk 11 above)
Comment (psmyrek) [reply docs/framework/guides/contributing/package-metadata.md]: Done.

File: docs/framework/guides/contributing/package-metadata.md (hunk 12 above)
Comment (psmyrek) [reply docs/framework/guides/contributing/package-metadata.md]: Done.

File: docs/framework/guides/contributing/package-metadata.md (hunk 13 above)
Comment (psmyrek) [reply docs/framework/guides/contributing/package-metadata.md]: Done.

File: docs/framework/guides/contributing/package-metadata.md (hunk 14 above)
Comment (psmyrek) [reply docs/framework/guides/contributing/package-metadata.md]: Changed "(...) that are created" to "(...) that are created or altered".

File: docs/framework/guides/creating-simple-plugin.md (hunk 15 above)
Comment (psmyrek) [reply docs/framework/guides/creating-simple-plugin.md]: Done.

**Comment by ma2ciek** [comment]
@godai78, could you do a final check of this PR?

File: docs/framework/guides/contributing/package-metadata.md [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+	* `elements` &ndash; HTML elements (a single one or an array of these) that are created or altered by the plugin. The pseudo-element `$block` indicates that a given plugin applies classes, styles or attributes (defined in appropriate properties) for all block elements.
+	* `classes` &ndash; CSS class names (a single one or an array of these) that may be applied to the HTML elements defined in the `elements` property.
+	* `styles` &ndash; Inline CSS styles (a single one or an array of these) that may be applied to the HTML elements defined in the `elements` property.
+	* `attributes` &ndash; HTML attributes (a single one or an array of these) other than `class` and `styles` (covered separately), that might be applied to the HTML elements defined in the `elements` property.
+	* `implements` &ndash; A name of an element or a pseudo-element which indicates that HTML elements defined in the `elements` property may contain classes, styles or attributes that are created by other plugins. For example, `implements` equal to `$block` means, that HTML elements may contain classes, styles or attributes that are defined by another plugin, which have `elements` equal to `$block`.
```
Comment (godai78) [inline docs/framework/guides/contributing/package-metadata.md]:
```suggestion
> For example, `implements` equal to `$block` means, that HTML elements may contain classes, styles or attributes that are defined by another plugin, which have `elements` equal to `$block`.
	
DO the classe, styles or attributes have `element`, or DOES the plugin have these? It is not clear now and if the latter, there needs to be 

> which has `elements` (...)	
```

Comment (godai78) [inline docs/framework/guides/contributing/package-metadata.md]: Here:  
  
\> classes, styles or attributes that are defined by another plugin, which have (...)  
  
if it's about the classes etc. or  
  
\> classes, styles or attributes that are defined by another plugin, which has (...)  
  
if it's about the plugin. This is not exactly clear.

Comment (psmyrek) [reply docs/framework/guides/contributing/package-metadata.md]: It is about the plugin, so I've just changed it to _"classes, styles or attributes that are defined by another plugin, which **has** (...)"_. Thanks.

Comment (godai78) [reply docs/framework/guides/contributing/package-metadata.md]: Cool, thanks!

**Comment by ma2ciek** [comment]
Thanks, @godai78 and @psmyrek!
